pkg/server/backuprestoreserver.go

```diff
@@ -401,25 +401,28 @@ func (b *BackupRestoreServer) runEtcdProbeLoopWithSnapshotter(ctx context.Contex
 							b.logger.Warnf("Snapshot lease update failed : %v", err)
 						}
 					}
-					if b.backoffConfig.Start {
+					if b.backoffConfig.Start && ssr.PrevFullSnapshotSucceed {
 						b.backoffConfig.ResetExponentialBackoff()
 					}
 				} else {
 					b.logger.Warnf("Failed to collect events for first delta snapshot(s): %v", err)
 				}
 			}
 
-			if !initialDeltaSnapshotTaken {
+			if !initialDeltaSnapshotTaken || !ssr.PrevFullSnapshotSucceed {
```

Comment:
> From my understanding, the suggested code currently makes use of the fact that:
ssr.takeFullSnapshot() errors, which causes
ssr.TakeFullSnapshotAndResetTimer() to error, which causes
ssr.snapshotEventHandler() to error while setting ssr.PrevFullSnapshotSucceed to false, which causes
ssr.Run() to error in b.runEtcdProbeLoopWithSnapshotter()
which then closes the snapshotter and restart it in b.runEtcdProbeLoopWithSnapshotter()
When the snapshotter is started again after being closed, it checks whether ssr.PrevFullSnapshotSucceed is set to false, and if it is, a new full snapshot is triggered.

yes, your understanding is correct.

> I think we can make use of the fact that in line 383, a call to IsFullSnapshotRequiredAtStartup()

I have deliberately avoided that to make reviewer life simple as this was urgent PR when it raised and I don't want reviewer to go inside `IsFullSnapshotRequiredAtStartup()` as logic inside that is little complex, and without context it is hard for reviewers to understand that.

Reply:
> We add an else block, and move all the code under if !initialDeltaSnapshotTaken to this else block, and get rid of the !initialDeltaSnapshotTaken condition.

no, that won't work. consider this scenario:
```
if (true){
unable to take delta as backup-restore unable to apply watch (as I told you yesterday)
}else{

}
```
this block of code won't work, what will work is:
```
if (true){
unable to take delta as backup-restore unable to apply watch (as I told you yesterday)
}
if (some condition to detect prev error) {

}
```

Reply:
I don't particularly think it is too complicated, we can modify `IsFullSnapshotRequiredAtStartup()` to include the case discussed.

> no, that won't work. consider this scenario:

You're right. The edge case that was discussed - etcd DB is compacted up to a revision which occurs after a revision that is being watched will cause the watch to fail; in which case a full snapshot is needed.

I feel the following is a simplification and would reduce the number of variables that one needs to keep track of mentally:

```go
fullRequired := false
if fullRequired = ssr.IsFullSnapshotRequiredAtStartup(); !fullRequired {
// trigger a delta since full is not needed
// ...
// some error occurs
// delta was not successful
fullRequired = true
}

if fullRequired {
// trigger a full snapshot since one of the possibilities is true
// 1. it is needed as delta snapshot could not be triggered for some reason
// 2. it was determined by IsFullSnapshotRequiredAtStartup()
// trigger the full
}
```

Reply:
you have just renamed `initialDeltaSnapshotTaken` to `fullRequired `, nothing else changes 😅

Reply:
> we don't need to maintain state in PrevFullSnapshotSucceed
no need to store this information again in PrevFullSnapshotSucceed

how's that possible, we still have to maintain `PrevFullSnapshotSucceed` whether you use inside `IsFullSnapshotRequiredAtStartup()` that's different thing dude. you can have to keep a track of success/failure of prev full snapshot. Please go through the code once again.

Reply:
- `TakeFullSnapshotAndResetTimer` errors due to some reason, which implies-
- `takeFullSnapshot` returns an error due to a failed snapshot-
  - `takeFullSnapshot` does not assign `ssr.PrevFullSnapshot = s` since `PrevFullSnapshot` is only updated when the full snapshot is *successful*.
  - `PrevFullSnapshot` will therefore contain the full snapshot that was triggered before the `timeWindow` that is present.

Thus, I think `PrevFullSnapshot` will only contain the previous successful full snapshot.

Now, if a particular full snapshot that was scheduled fails, then we can say that time duration since it was created would be more than the `timeWindow`.

```go
func (ssr *Snapshotter) IsFullSnapshotRequiredAtStartup(timeWindow float64) bool {
	if ssr.PrevFullSnapshot == nil || ssr.PrevFullSnapshot.IsFinal || time.Since(ssr.PrevFullSnapshot.CreatedOn).Hours() > timeWindow {
		return true
	}

	if !ssr.WasScheduledFullSnapshotMissed(timeWindow) {
		return false
	}
	return ssr.IsNextFullSnapshotBeyondTimeWindow(timeWindow)
}
```

`time.Since(ssr.PrevFullSnapshot.CreatedOn).Hours() > timeWindow` would be `true`, and  thereby return `true`.

Can you help me understand where I'm wrong?

Reply:
`PrevFullSnapshotSucceed ` is being in snapshotEventHandler to capture the state.

Reply:
> takeFullSnapshot does not assign ssr.PrevFullSnapshot = s since PrevFullSnapshot is only updated when the full snapshot is successful.

but still be holding the previous state of previous successful state, if you want to reuse this then before returning the error you have to make it nil

Reply:
> PrevFullSnapshot will therefore contain the full snapshot that was triggered before the timeWindow that is present.

yes, that should be the case. I have already mentioned this in slack thread:
```
While making changes for this, I had taken a look at current codebase and I found that if today full-snapshot fails then backup-restore is smart enough to trigger a full-snapshot again.
```

Reply:
Thanks for your explanation!